frontend/src/metabase/sharing/components/EmailAttachmentPicker.jsx

```diff
@@ -20,140 +23,163 @@ export default class EmailAttachmentPicker extends Component {
     setPulse: PropTypes.func.isRequired,
   };
 
-  DEFAULT_ATTACHMENT_TYPE = "csv";
+  componentDidMount() {
+    this.setState(this.calculateStateFromCards());
+  }
 
-  updateStateFromCards() {
-    const { cards } = this.props;
-    const { attachmentType } = this.state;
+  componentDidUpdate() {
+    const newState = this.calculateStateFromCards();
+    
+    newState.isEnabled = newState.isEnabled || this.state.isEnabled;
+    if(newState.selectedCardIds.size == 0) newState.selectedAttachmentType = this.state.selectedAttachmentType;
+    
+    if(!this.shouldUpdateState(newState, this.state)) {
+      this.setState(newState);
+    }
+  }
 
+  calculateStateFromCards() {
+    const { cards } = this.props;
     const selectedCards = cards.filter(card => {
       return card.include_csv || card.include_xls;
     });
 
-    if (!attachmentType && selectedCards.length > 0) {
-      this.setState({
-        attachmentType: this.attachmentTypeFor(selectedCards),
-        selectedCardIds: new Set(selectedCards.map(card => card.id)),
-      });
-      this.updatePulse();
+    return {
+      isEnabled: selectedCards.length > 0,
+      selectedAttachmentType: this.attachmentTypeFor(selectedCards) || this.DEFAULT_ATTACHMENT_TYPE,
+      selectedCardIds: new Set(selectedCards.map(card => card.id)),
     }
   }
 
-  componentDidMount() {
-    this.updateStateFromCards();
+  shouldUpdateState(newState, currentState) {
+    return (currentState.isEnabled || !newState.isEnabled) &&
+      newState.selectedAttachmentType === currentState.selectedAttachmentType &&
+      this.setsEqual(newState.selectedCardIds, currentState.selectedCardIds);
   }
 
-  componentDidUpdate() {
-    this.updateStateFromCards();
+  setsEqual(as, bs) {
+    if (as.size !== bs.size) {
+      return false;
+    }
+    for (let a of as) {
+      if (!bs.has(a)) {
+        return false;
+      }
+    }
+    return true;
+  }
```

Comment:
I was a little shocked I had to write this, but Stack Overflow told me to :shrug:

Reply:
Does `_` or maybe `icepick` have some sort of function that we could use instead of writing our own here?